File: ruby/vending_machine/vending_machine.rb
require_relative 'juice'
require_relative 'suica'

# VendingMachineクラスの定義
class VendingMachine
  attr_reader :sell_amt

  def initialize
    @products = []
    5.times do
      @products << Juice.new('ペプシ', 150)
      @products << Juice.new('モンスター', 230)
      @products << Juice.new('いろはす', 120)
    end

    # 売上
    @sell_amt = 0
  end

  # 飲み物購入
  def buy(name, suica)
    err_msg = can_buy_check(name, suica)
    raise err_msg unless err_msg == ''

    idx = @products.index { |val| val.name == name }

    @sell_amt += @products[idx].price
    suica.pay(@products[idx].price)
    @products.delete_at(idx)
  end

  # 在庫リスト
  def stock_list
    ret_list(:stock)
  end

  # 購入可能リスト
  def can_buy_list
    ret_list(:can_buy)
  end

  # 在庫補充
  def restock_drink(name, price, qty)
    qty.times do
      @products << Juice.new(name, price)
    end
  end

  private

  def ret_list(param)
    drinks = @products.map(&:name).uniq
    list = []
    if param == :stock
      drinks.each do |drink|
        qty = @products.select { |n| n.name == drink }.count
        list << "#{drink} : #{qty}本"
      end
    elsif param == :can_buy
      list = drinks
    end
    list.join(',')
  end

  def can_buy_check(name, suica)
    err_list = []
    err_list << '在庫がないため購入できません。' unless @products.map(&:name).uniq.include?(name)

    @products.each do |item|
      err_list << '残高が足りないため購入できません。' if item.name == name && item.price > suica.amt
    end
    err_list.join('\\n')
  end
end

File: python/vending_machine/vending_machine.py
from drink import Drink
from suica import Suica


class VendingMachine:
    def __init__(self):
        self._stocks = []
        for _ in range(5):
            self._stocks.append(Drink("ペプシ", 150))
            self._stocks.append(Drink("モンスター", 230))
            self._stocks.append(Drink("いろはす", 120))
        self._amount = 0

    @property
    def amount(self):
        return self._amount

    def get_stock_info(self, info_style):
        """
        ストックが1本以上ある購入可能なドリンク情報または本数まで含んだストックの情報を返す

        Args:
          info_style (str): 情報の種別 'stock'または'can_buy'

        Raises:
          ValueError: 金額が100円以下の場合に発生
        """
        drink_list = []
        stock_set = set([drink.name for drink in self._stocks])

        if info_style == "stock":
            for drink_name in stock_set:
                qty = len([drink for drink in self._stocks if drink.name == drink_name])
                drink_list.append(f"{drink_name} : {qty}本")
        elif info_style == "can_buy":
            drink_list = list(stock_set)
        return ", ".join(drink_list)

    def buy(self, suica, drink_name):
        # 購入可能かのチェック
        self.can_buy_check(suica, drink_name)

        stock_idx = [stock.name for stock in self._stocks].index(drink_name)
        price = self._stocks[stock_idx].price

        suica.pay(price)
        self._amount += price
        del self._stocks[stock_idx]

    def can_buy_check(self, suica, drink_name):
        """
        ドリンクが購入可能かチェックする。
          在庫が存在するか
          残金不足でないか

        Args:
          suica (instance): Suicaクラスのインスタンス
          drink_name (str): 購入するdrinkの名称

        Raises:
          Exception: 在庫不足の場合または残金不足の場合に発生
        """
        buy_drink_info = next(
            (stock for stock in self._stocks if stock.name == drink_name), None
        )

        if buy_drink_info is None:
            raise Exception("在庫がないため購入できません。")

        if suica.deposit < buy_drink_info.price:
            raise Exception("残金不足のため購入できません。")

    def refill(self, name, price):
        """
        self._stocksにdrinkを追加する

        Args:
          name (str): 追加するdrinkの名称

          price (sint): 追加するdrinkの金額

        Raises:
        """
        self._stocks.append(Drink(name, price))


if __name__ == "__main__":
    my_machine = VendingMachine()
    abc = my_machine.get_stock()
    print(abc)
